### PR TITLE
Add uniqueness (`&mut`-ability) to `NodeInfo`

### PR DESCRIPTION
**Pull request by boyland-pf**
Finished generating info for the uniqueness "query". 10 unit tests. The one thing I didn't include is formatting because that would require a new snapshot.

### CODE REVIEW

File: pdg/src/info.rs
Comment (aneksteind) [inline pdg/src/info.rs:18]: documenting `unique` and what it means would be helpful

Comment (boyland-pf) [reply pdg/src/info.rs:18]: what if I add in the parenthetical
`as well as its ability to be used as a `&mut` (signified by 'unique')`?

File: pdg/src/info.rs
Comment (aneksteind) [inline pdg/src/info.rs:101]: this for loop can go above the first, the other can just push to the existing vector without `or_insert_with` or `new`

File: pdg/src/info.rs
Comment (aneksteind) [inline pdg/src/info.rs:110]:
```suggestion
/// Given a list of nodes of the same parent and information about them,
/// determines if any have conflicts with any of the others.
/// Children which are not a field cannot be live at the same time as any other child.
/// Children which are a field cannot be live at the same time as any other one of the same field.
```

File: pdg/src/info.rs
Comment (aneksteind) [inline pdg/src/info.rs:123]: have you run `rustfmt`?

File: pdg/src/info.rs
Comment (aneksteind) [inline pdg/src/info.rs:122]: i think this deserves a comment explaining what you're matching for here and what makes it a conflict

File: pdg/src/info.rs
Comment (aneksteind) [reply pdg/src/info.rs:18]: i think it should be added above the attribute itself, not in the general struct documentation

File: pdg/src/info.rs
Comment (boyland-pf) [reply pdg/src/info.rs:122]: This part isn't changed by it.

File: pdg/src/info.rs
Comment (aneksteind) [inline pdg/src/info.rs:102]: i think you can `map` here instead of unwrap

File: pdg/src/info.rs
Comment (aneksteind) [inline pdg/src/info.rs:143]: does this imply that the whole connected tree is non-unique if any nodes contained in it are?

File: pdg/src/info.rs
Comment (aneksteind) [reply pdg/src/info.rs:122]: i would add some spacing after the comments and see if that changes, makes things more readable IMO

File: pdg/src/info.rs
Comment (boyland-pf) [reply pdg/src/info.rs:123]: I've added a comment describing what the two cases are

File: pdg/src/info.rs
Comment (boyland-pf) [reply pdg/src/info.rs:145]: There are test cases in which the other direction doesn't hold - if there's a conflict between A's child B and A's child C, A can't be unique, but its parent Z can still be.

File: pdg/src/info.rs
Comment (boyland-pf) [reply pdg/src/info.rs:102]: we're guaranteed the unwrap will work, so isn't unwrap better here because it's simpler?

File: pdg/src/info.rs
Comment (aneksteind) [reply pdg/src/info.rs:145]: the reason I asked is because I can't tell whether or not the order of traversal in nodes matters or not, but it sounds like it doesn't because it's strictly limited to immediate parent and children. is that right?

File: pdg/src/info.rs
Comment (aneksteind) [reply pdg/src/info.rs:102]: i think `map` is better because it decouples this line of code from the knowledge of what the code above it does

File: pdg/src/info.rs
Comment (aneksteind) [inline pdg/src/info.rs:349]: i'm trying to make sense of these, why should the ancestors of `a` be non-unique if they're written to?

File: pdg/src/info.rs
Comment (aneksteind) [inline pdg/src/info.rs:366]: spaces after `//` would help readability

File: pdg/src/info.rs
Comment (kkysen) [inline pdg/src/info.rs:94]: What's `par`?

File: pdg/src/info.rs
Comment (kkysen) [reply pdg/src/info.rs:102]: How would `.map` be used here?

File: pdg/src/info.rs
Comment (boyland-pf) [reply pdg/src/info.rs:94]: each possible parent

File: pdg/src/info.rs
Comment (aneksteind) [reply pdg/src/info.rs:349]: after our conversation offline, the test case makes sense to me

File: pdg/src/info.rs
Comment (aneksteind) [reply pdg/src/info.rs:509]: and please check the rest for parity

File: pdg/src/info.rs
Comment (aneksteind) [inline pdg/src/info.rs:513]: i think it would be helpful to have variables such as
```rust
let (x, y, z) = (0_u32, 1_u32, 2_u32);
```
to correspond to the comments more accurately

File: pdg/src/info.rs
Comment (aneksteind) [inline pdg/src/info.rs:499]: thoughts on interleaving these comments above each respective line, rather than have them all at the top?

File: pdg/src/info.rs
Comment (aneksteind) [reply pdg/src/info.rs:102]: would `m.get_mut(&par).map(|ref mut v| v.push(chi); v)` or not work? but on second thought maybe `unwrap` is fine because the map option relies on side affects

Comment (kkysen) [reply pdg/src/info.rs:102]: Yeah, `map` is generally used for mapping to a value, not running a side effect.  The equivalent to the `map` but more idiomatic would be an `if let Some`.  But we also want to assert here that it's `Some`, right?  At the very least, `unwrap` serves as documentation that it will always be `Some`.

File: pdg/src/info.rs
Comment (kkysen) [reply pdg/src/info.rs:499]: I think it should be in doc comments for the function so that it's run and tested.  It's easy to make a small mistake as is.

File: pdg/src/info.rs
Comment (boyland-pf) [reply pdg/src/info.rs:145]: Order of traversal does matter, because any descendant of a non-unique node is non-unique. My assumption which I believe is listed in the comments is that the nodes are ordered.

File: pdg/src/info.rs
Comment (aneksteind) [reply pdg/src/info.rs:102]: that sounds reasonable

File: pdg/src/info.rs
Comment (aneksteind) [reply pdg/src/info.rs]: i don't see how these comments can be tested. considering that this function is the test itself, what do you envision?

Comment (kkysen) [reply pdg/src/info.rs]: I mean we can test that they're well-formed Rust.

Comment (boyland-pf) [reply pdg/src/info.rs]: I believe I have tested them all, but that would make it easier to be sure, and also if there's changes...

Comment (aneksteind) [reply pdg/src/info.rs]: my primary concern is attributing lines here to lines of test code below. i still would consider interleaving them for that purpose because i think having them spatially local helps with that. testing that it's well formed rust seems orthogonal to that

File: pdg/src/info.rs
Comment (boyland-pf) [reply pdg/src/info.rs:124]: I've done that, but it hasn't added any newlines, if you want me to add one I'm not sure where to put it.

Comment (aneksteind) [reply pdg/src/info.rs:124]: it's fine, not the biggest deal

File: pdg/src/info.rs
Comment (aneksteind) [inline pdg/src/info.rs:39]:
```suggestion
    /// Initializes a [`FlowInfo`] based on a node's [`NodeKind`]
```

File: pdg/src/info.rs
Comment (aneksteind) [inline pdg/src/info.rs:122]:
```suggestion
        // if the first below matches, then two siblings, neither a field, conflict
        // if the second matches, then two siblings of the same field conflict
```

File: pdg/src/info.rs
Comment (aneksteind) [inline pdg/src/info.rs:145]: I think we talked about adding comments to explain the relationship between uniqueness and parentage, etc.

File: pdg/src/info.rs
Comment (kkysen) [inline pdg/src/info.rs:41]:
```suggestion
    /// Initializes a [`FlowInfo`] based on a [`Node`]'s [`NodeKind`].
```

File: pdg/src/info.rs
Comment (kkysen) [inline pdg/src/info.rs:109]:
```suggestion
/// Given a list of [`Node`]s of the same parent and information about them,
```

File: pdg/src/info.rs
Comment (kkysen) [inline pdg/src/info.rs:385]:
```suggestion
    /// let mut a = Point {x: 0, y: 0};
```

File: pdg/src/info.rs
Comment (kkysen) [inline pdg/src/info.rs:400]:
```suggestion
        // let mut a = Point {x: 0, y: 0};
```

File: pdg/src/info.rs
Comment (kkysen) [inline pdg/src/info.rs:422]:
```suggestion
    /// let mut a = Point {x: 0, y: 0};
```

File: pdg/src/info.rs
Comment (kkysen) [inline pdg/src/info.rs:443]:
```suggestion
        // let mut a = Point {x: 0, y: 0};
```

File: pdg/src/info.rs
Comment (kkysen) [inline pdg/src/info.rs:472]:
```suggestion
    /// let mut a = Point {x: 0, y: 0};
```

File: pdg/src/info.rs
Comment (kkysen) [inline pdg/src/info.rs:487]:
```suggestion
        // let mut a = Point {x: 0, y: 0};
```

File: pdg/src/info.rs
Comment (kkysen) [inline pdg/src/info.rs:508]:
```suggestion
    /// let mut a = Point {x: 0, y: 0};
```

File: pdg/src/info.rs
Comment (kkysen) [inline pdg/src/info.rs:524]:
```suggestion
        // let mut a = Point {x: 0, y: 0};
```

**Review by aneksteind — APPROVED** [review]
this LGTM after @spernsteiner confirms that it's functionally correct

**Review by kkysen — DISMISSED** [review]
I'd like to merge #666 into this before merging, as well as another PR for cleaning up some stuff.  I'll try to get that next PR done soon, so if I take longer, we can merge this PR and I'll open those subsequent PRs later.  #666 should merge first, though.

**Comment by spernsteiner** [comment]
Sorry for the delay on this.  I poked at this code a bit and found some bugs.  I've pushed a commit with some additional tests to demonstrate.

I think the general issue is that `check_children_conflict` only looks at the immediate children of a node, but in fact it needs to look at entire chains of `Field` nodes to figure out what really conflicts.  `Field(0)` conflicts with `Field(0)`, but `Field(0) -> Field(0)` does not conflict with `Field(0) -> Field(1)`.  Concretely, you can have `let p = &mut x.0.0; let q = &mut x.0.1;` and overlapping uses of `p` and `q` do not cause a borrow checker error (the test `nested_field_no_conflict` shows an example of this).

**Comment by boyland-pf** [comment]
Thanks, that example makes sense. Can you explain, though, why "offset_field_no_conflict" should actually find a conflict? Rust is willing to compile it, at least on my machine.

File: pdg/src/info.rs
Comment (kkysen) [reply pdg/src/info.rs:94]: `par` is confusing.  Can we just call it `parent`?  `par` is more often `parallel`.  Same with `chi` and `child`.

File: pdg/src/info.rs
Comment (kkysen) [inline pdg/src/info.rs:107]: Why do we need two passes over the nodes?

**Comment by kkysen** [comment]
Can all the `HashMap<NodeId, T>`s be `IndexVec<NodeId, T>`s?

File: pdg/src/info.rs
Comment (kkysen) [reply pdg/src/info.rs:107]: Hmm, with an `InddxVec` (basically just a `Vec`, the first pass should just be a `memset`, so I'll wait until we convert the `HashMap`s to `IndexVec`s.

**Comment by boyland-pf** [comment]
On the subject of a[1].0 not being able to be live at the same time as a[1].1 (even though rustc allows it), I’m assuming that the modification needed to the algorithm output is that any descendent of an offset is required to not be live at the same time as any other descendent of an offset of the same variable.

If so, will push a fix in a couple hours both for this and the more obvious field one.

**Comment by spernsteiner** [comment]
My original thinking was that offset nodes don't require any special handling - only fields require you to look through several levels before deciding whether there's a conflict.  But I haven't put much thought into the details, so that could be wrong.

**Comment by boyland-pf** [comment]
The code should now be correct, though somewhat messier again. In particular, it is currently somewhat inefficient; however, the inefficiency is entirely within a hashmap that can be refactored into a trie, which would be efficient.

Some test cases were changed because the code that was in their comments had `let x = a.0;` as merely a field node instead of a field followed by a copy as it should be. I added a complement to the `nested_field_no_conflict` test case showing the case where storing an intermediate field in a variable changes the &mut-ability.

File: pdg/src/info.rs
Comment (kkysen) [inline pdg/src/info.rs:842]:
```suggestion
    /// ```
```

File: pdg/src/info.rs
Comment (kkysen) [inline pdg/src/info.rs:149]: Is this checking if one of `fs1` and `fs2` is a prefix of the other?

Comment (boyland-pf) [reply pdg/src/info.rs:149]: Yes, and that's a job only efficient to do with prefix trees or some other data structure intended for it.

Comment (kkysen) [reply pdg/src/info.rs:149]: This doesn't seem very inefficient?  It should be able to be compiled to just a `memcmp`.  And if performance is important, we can use `tinyvec` for the `Vec<Field>`s and the `memcmp` should be short and cache-resident, usually.  There shouldn't be very many `Field`s in a `Vec<Field>`, right?

Also, I can make this more idiomatic.

Comment (boyland-pf) [reply pdg/src/info.rs:149]: The inefficient part isn't this, but having to run this for every single child. Though I suppose this is only inefficient in theory, because maybe there won't ever be any nodes with more than a few dozen immediate children.

Comment (kkysen) [reply pdg/src/info.rs:149]: Ah, got it.  Well, let's save any optimizations until later when we know it's actually a bottleneck, or if a better trie solution is simpler.

**Comment by boyland-pf** [comment]
Outside of an efficiency issue that only appears if no maximum degree of field children is assumed, this appears done to me.

I would appreciate in particular my change to the `nested_field_no_conflict` to be looked at, standing in for the changes made throughout to bring the test pdg in line with what the pdg of the example rust code should actually be. The contrast between `nested_field_no_conflict` and `diff_field_conflict` may be interesting.

**Comment by boyland-pf** [comment]
@spernsteiner Could you take a look at this?

**Comment by spernsteiner** [comment]
>I would appreciate in particular my change to the `nested_field_no_conflict` test to be looked at, standing in for the changes made throughout to bring the test pdg in line with what the pdg of the example rust code should actually be.

The only change in `nested_field_no_conflict` is the addition of the two `Copy` nodes, right?  That seems fine.

>The contrast between `nested_field_no_conflict` and `diff_field_conflict` may be interesting.

That is interesting!  I hadn't considered that there would be a difference between `&mut a.0.0` and `&mut (*&mut a.0).0`.  I was thinking of `Copy` as a no-op that could potentially be elided if it didn't impact the graph structure, but I see now that's not the case.

**Review by spernsteiner — APPROVED** [review]
I haven't looked at the implementation, but the changes to the tests look good to me.

File: pdg/src/info.rs
Comment (spernsteiner) [reply pdg/src/info.rs:511]: There are similar issues on several other tests.  In general, accessing a field behind a reference should be either `(*a).y` or just `a.y`, rather than `*(a.y)` (which indicates dereferencing a reference stored in `a.y` instead).

**Pull request merged**
